File: src/serial/record_encoder.cc
// Copyright (c) 2023 dingodb.com, Inc. All Rights Reserved
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "record_encoder.h"

#include <sys/types.h>

#include <memory>
#include <string>

#include "proto/common.pb.h"
#include "serial/keyvalue.h"  // IWYU pragma: keep

namespace dingodb {

RecordEncoder::RecordEncoder(int schema_version, std::shared_ptr<std::vector<std::shared_ptr<BaseSchema>>> schemas,
                             long common_id) {
  this->le_ = IsLE();
  Init(schema_version, schemas, common_id);
}

RecordEncoder::RecordEncoder(int schema_version, std::shared_ptr<std::vector<std::shared_ptr<BaseSchema>>> schemas,
                             long common_id, bool le) {
  this->le_ = le;
  Init(schema_version, schemas, common_id);
}

void RecordEncoder::Init(int schema_version, std::shared_ptr<std::vector<std::shared_ptr<BaseSchema>>> schemas,
                         long common_id) {
  this->schema_version_ = schema_version;
  FormatSchema(schemas, this->le_);
  this->schemas_ = schemas;
  this->common_id_ = common_id;
  int32_t* size = GetApproPerRecordSize(schemas);
  this->key_buf_size_ = size[0];
  this->value_buf_size_ = size[1];
  delete[] size;
}

void RecordEncoder::EncodePrefix(Buf& buf, char prefix) const {
  buf.Write(prefix);
  buf.WriteLong(common_id_);
}

void RecordEncoder::EncodeReverseTag(Buf& buf) const {
  buf.ReverseWrite(codec_version_);
  buf.ReverseWrite(0);
  buf.ReverseWrite(0);
  buf.ReverseWrite(0);
}

void RecordEncoder::EncodeSchemaVersion(Buf& buf) const { buf.WriteInt(schema_version_); }

int RecordEncoder::Encode(char prefix, const std::vector<std::any>& record, std::string& key, std::string& value) {
  int ret = EncodeKey(prefix, record, key);
  if (ret < 0) {
    return ret;
  }
  ret = EncodeValue(record, value);
  if (ret < 0) {
    return ret;
  }
  return 0;
}

int RecordEncoder::Encode(char prefix, const std::vector<std::any>& record, pb::common::KeyValue& key_value) {
  int ret = EncodeKey(prefix, record, *key_value.mutable_key());
  if (ret < 0) {
    return ret;
  }
  ret = EncodeValue(record, *key_value.mutable_value());
  if (ret < 0) {
    return ret;
  }
  return 0;
}

int RecordEncoder::EncodeKey(char prefix, const std::vector<std::any>& record, std::string& output) {
  Buf buf(key_buf_size_, this->le_);
  // |namespace|id| ... |tag|
  buf.EnsureRemainder(13);
  EncodePrefix(buf, prefix);
  EncodeReverseTag(buf);
  int index = 0;
  for (const auto& bs : *schemas_) {
    if (bs) {
      BaseSchema::Type type = bs->GetType();
      switch (type) {
        case BaseSchema::kBool: {
          auto bos = std::dynamic_pointer_cast<DingoSchema<std::optional<bool>>>(bs);
          if (bos->IsKey()) {
            bos->EncodeKey(&buf, std::any_cast<std::optional<bool>>(record.at(index)));
          }
          break;
        }
        case BaseSchema::kInteger: {
          auto is = std::dynamic_pointer_cast<DingoSchema<std::optional<int32_t>>>(bs);
          if (is->IsKey()) {
            is->EncodeKey(&buf, std::any_cast<std::optional<int32_t>>(record.at(index)));
          }
          break;
        }
        case BaseSchema::kFloat: {
          auto fs = std::dynamic_pointer_cast<DingoSchema<std::optional<float>>>(bs);
          if (fs->IsKey()) {
            fs->EncodeKey(&buf, std::any_cast<std::optional<float>>(record.at(index)));
          }
          break;
        }
        case BaseSchema::kLong: {
          auto ls = std::dynamic_pointer_cast<DingoSchema<std::optional<int64_t>>>(bs);
          if (ls->IsKey()) {
            ls->EncodeKey(&buf, std::any_cast<std::optional<int64_t>>(record.at(index)));
          }
          break;
        }
        case BaseSchema::kDouble: {
          auto ds = std::dynamic_pointer_cast<DingoSchema<std::optional<double>>>(bs);
          if (ds->IsKey()) {
            ds->EncodeKey(&buf, std::any_cast<std::optional<double>>(record.at(index)));
          }
          break;
        }
        case BaseSchema::kString: {
          auto ss = std::dynamic_pointer_cast<DingoSchema<std::optional<std::shared_ptr<std::string>>>>(bs);
          if (ss->IsKey()) {
            ss->EncodeKey(&buf, std::any_cast<std::optional<std::shared_ptr<std::string>>>(record.at(index)));
          }
          break;
        }
        default: {
          break;
        }
      }
    }
    index++;
  }

  return buf.GetBytes(output);
}

int RecordEncoder::EncodeValue(const std::vector<std::any>& record, std::string& output) {
  Buf buf(value_buf_size_, this->le_);
  buf.EnsureRemainder(4);
  EncodeSchemaVersion(buf);

  for (const auto& bs : *schemas_) {
    if (bs) {
      BaseSchema::Type type = bs->GetType();
      switch (type) {
        case BaseSchema::kBool: {
          auto bos = std::dynamic_pointer_cast<DingoSchema<std::optional<bool>>>(bs);
          if (!bos->IsKey()) {
            bos->EncodeValue(&buf, std::any_cast<std::optional<bool>>(record.at(bos->GetIndex())));
          }
          break;
        }
        case BaseSchema::kInteger: {
          auto is = std::dynamic_pointer_cast<DingoSchema<std::optional<int32_t>>>(bs);
          if (!is->IsKey()) {
            is->EncodeValue(&buf, std::any_cast<std::optional<int32_t>>(record.at(is->GetIndex())));
          }
          break;
        }
        case BaseSchema::kFloat: {
          auto fs = std::dynamic_pointer_cast<DingoSchema<std::optional<float>>>(bs);
          if (!fs->IsKey()) {
            fs->EncodeValue(&buf, std::any_cast<std::optional<float>>(record.at(fs->GetIndex())));
          }
          break;
        }
        case BaseSchema::kLong: {
          auto ls = std::dynamic_pointer_cast<DingoSchema<std::optional<int64_t>>>(bs);
          if (!ls->IsKey()) {
            ls->EncodeValue(&buf, std::any_cast<std::optional<int64_t>>(record.at(ls->GetIndex())));
          }
          break;
        }
        case BaseSchema::kDouble: {
          auto ds = std::dynamic_pointer_cast<DingoSchema<std::optional<double>>>(bs);
          if (!ds->IsKey()) {
            ds->EncodeValue(&buf, std::any_cast<std::optional<double>>(record.at(ds->GetIndex())));
          }
          break;
        }
        case BaseSchema::kString: {
          auto ss = std::dynamic_pointer_cast<DingoSchema<std::optional<std::shared_ptr<std::string>>>>(bs);
          if (!ss->IsKey()) {
            ss->EncodeValue(&buf,
                            std::any_cast<std::optional<std::shared_ptr<std::string>>>(record.at(ss->GetIndex())));
          }
          break;
        }
        case BaseSchema::kBoolList: {
          auto ss = std::dynamic_pointer_cast<DingoSchema<std::optional<std::shared_ptr<std::vector<bool>>>>>(bs);
          if (!ss->IsKey()) {
            ss->EncodeValue(
                &buf, std::any_cast<std::optional<std::shared_ptr<std::vector<bool>>>>(record.at(ss->GetIndex())));
          }
          break;
        }
        case BaseSchema::kStringList: {
          auto ss =
              std::dynamic_pointer_cast<DingoSchema<std::optional<std::shared_ptr<std::vector<std::string>>>>>(bs);
          if (!ss->IsKey()) {
            ss->EncodeValue(&buf, std::any_cast<std::optional<std::shared_ptr<std::vector<std::string>>>>(
                                      record.at(ss->GetIndex())));
          }
          break;
        }
        case BaseSchema::kDoubleList: {
          auto ss = std::dynamic_pointer_cast<DingoSchema<std::optional<std::shared_ptr<std::vector<double>>>>>(bs);
          if (!ss->IsKey()) {
            ss->EncodeValue(
                &buf, std::any_cast<std::optional<std::shared_ptr<std::vector<double>>>>(record.at(ss->GetIndex())));
          }
          break;
        }
        case BaseSchema::kFloatList: {
          auto ss = std::dynamic_pointer_cast<DingoSchema<std::optional<std::shared_ptr<std::vector<float>>>>>(bs);
          if (!ss->IsKey()) {
            ss->EncodeValue(
                &buf, std::any_cast<std::optional<std::shared_ptr<std::vector<float>>>>(record.at(ss->GetIndex())));
          }
          break;
        }
        case BaseSchema::kIntegerList: {
          auto ss = std::dynamic_pointer_cast<DingoSchema<std::optional<std::shared_ptr<std::vector<int32_t>>>>>(bs);
          if (!ss->IsKey()) {
            ss->EncodeValue(
                &buf, std::any_cast<std::optional<std::shared_ptr<std::vector<int32_t>>>>(record.at(ss->GetIndex())));
          }
          break;
        }
        case BaseSchema::kLongList: {
          auto ss = std::dynamic_pointer_cast<DingoSchema<std::optional<std::shared_ptr<std::vector<int64_t>>>>>(bs);
          if (!ss->IsKey()) {
            ss->EncodeValue(
                &buf, std::any_cast<std::optional<std::shared_ptr<std::vector<int64_t>>>>(record.at(ss->GetIndex())));
          }
          break;
        }
        default: {
          break;
        }
      }
    }
  }

  return buf.GetBytes(output);
}

int RecordEncoder::EncodeKeyPrefix(char prefix, const std::vector<std::any>& record, int column_count,
                                   std::string& output) {
  Buf buf(key_buf_size_, this->le_);
  buf.EnsureRemainder(9);
  EncodePrefix(buf, prefix);
  for (const auto& bs : *schemas_) {
    if (bs) {
      BaseSchema::Type type = bs->GetType();
      switch (type) {
        case BaseSchema::kBool: {
          auto bos = std::dynamic_pointer_cast<DingoSchema<std::optional<bool>>>(bs);
          if (bos->IsKey()) {
            bos->EncodeKeyPrefix(&buf, std::any_cast<std::optional<bool>>(record.at(bos->GetIndex())));
          }
          break;
        }
        case BaseSchema::kInteger: {
          auto is = std::dynamic_pointer_cast<DingoSchema<std::optional<int32_t>>>(bs);
          if (is->IsKey()) {
            is->EncodeKeyPrefix(&buf, std::any_cast<std::optional<int32_t>>(record.at(is->GetIndex())));
          }
          break;
        }
        case BaseSchema::kFloat: {
          auto fs = std::dynamic_pointer_cast<DingoSchema<std::optional<float>>>(bs);
          if (fs->IsKey()) {
            fs->EncodeKeyPrefix(&buf, std::any_cast<std::optional<float>>(record.at(fs->GetIndex())));
          }
          break;
        }
        case BaseSchema::kLong: {
          auto ls = std::dynamic_pointer_cast<DingoSchema<std::optional<int64_t>>>(bs);
          if (ls->IsKey()) {
            ls->EncodeKeyPrefix(&buf, std::any_cast<std::optional<int64_t>>(record.at(ls->GetIndex())));
          }
          break;
        }
        case BaseSchema::kDouble: {
          auto ds = std::dynamic_pointer_cast<DingoSchema<std::optional<double>>>(bs);
          if (ds->IsKey()) {
            ds->EncodeKeyPrefix(&buf, std::any_cast<std::optional<double>>(record.at(ds->GetIndex())));
          }
          break;
        }
        case BaseSchema::kString: {
          auto ss = std::dynamic_pointer_cast<DingoSchema<std::optional<std::shared_ptr<std::string>>>>(bs);
          if (ss->IsKey()) {
            ss->EncodeKeyPrefix(&buf,
                                std::any_cast<std::optional<std::shared_ptr<std::string>>>(record.at(ss->GetIndex())));
          }
          break;
        }
        default: {
          break;
        }
      }
    }

    if (--column_count <= 0) {
      break;
    }
  }

  return buf.GetBytes(output);
}

int RecordEncoder::EncodeMaxKeyPrefix(char prefix, std::string& output) const {
  if (common_id_ == INT64_MAX) {
    return -1;
  }

  Buf buf(key_buf_size_, this->le_);
  buf.EnsureRemainder(9);
  buf.Write(prefix);
  buf.WriteLong(common_id_ + 1);
  return buf.GetBytes(output);
}

int RecordEncoder::EncodeMinKeyPrefix(char prefix, std::string& output) const {
  Buf buf(key_buf_size_, this->le_);
  buf.EnsureRemainder(9);
  buf.Write(prefix);
  buf.WriteLong(common_id_);

  return buf.GetBytes(output);
}

}  // namespace dingodb

File: java/dingo-sdk/src/main/java/io/dingodb/sdk/common/serial/RecordEncoder.java
/*
 * Copyright 2021 DataCanvas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.dingodb.sdk.common.serial;

import io.dingodb.sdk.common.KeyValue;
import io.dingodb.sdk.common.serial.schema.DingoSchema;
import lombok.Setter;

import java.util.Arrays;
import java.util.List;

public class RecordEncoder {
    private final int schemaVersion;
    private List<DingoSchema> schemas;
    private final long id;
    private int keyBufSize;
    private int valueBufSize;
    private int valueColumnCount;

    @Setter
    private int codecVersion;

    private byte[] encodeValueV1(Object[] record) {
        Buf valueBuf = new BufImpl(valueBufSize);
        encodeSchemaVersion(valueBuf);
        for (DingoSchema schema : schemas) {
            if (!schema.isKey()) {
                schema.encodeValue(valueBuf, record[schema.getIndex()]);
            }
        }
        return valueBuf.getBytes();
    }

    private byte[] encodeValueV2(Object[] record) {
        int cntNotNullCols = 0;
        int cntNullCols = 0;
        int totalLength = 0;

        int cntNotNullPos = 4;  //The first 4 bytes is for schema version.
        int cntNullPos = cntNotNullPos + 2;
        int idPos = cntNullPos + 2;
        int offsetPos = idPos + Config.idUnit * this.valueColumnCount;
        int dataPos = offsetPos + Config.offsetUnit * this.valueColumnCount;
        totalLength = 8;

        Buf valueBuf = new BufImpl(dataPos + valueBufSize, dataPos);
        encodeSchemaVersion(0, valueBuf);
        for (DingoSchema schema : schemas) {

            if (!schema.isKey()) {
                int index = schema.getIndex();
                Object column = record[index];

                if (column == null) {
                    cntNullCols++;

                    //Write id.
                    valueBuf.writeShort(idPos, (short) index);
                    idPos += 2;

                    //write offset.
                    valueBuf.writeInt(offsetPos, -1);
                    offsetPos += 4;
                } else {
                    cntNotNullCols++;

                    //write id.
                    valueBuf.writeShort(idPos, (short) index);
                    idPos += 2;

                    //write offset.
                    valueBuf.writeInt(offsetPos, dataPos);
                    offsetPos += 4;

                    //write data.
                    dataPos += schema.encodeValueV2(valueBuf, record[schema.getIndex()]);
                }
            }
        }

        valueBuf.writeShort(cntNotNullPos, (short) cntNotNullCols);
        valueBuf.writeShort(cntNullPos, (short) cntNullCols);

        return valueBuf.getBytes();
    }

    private int getValueColumnCount() {
        int count = 0;
        for ( DingoSchema schema : schemas ) {
            if (!schema.isKey()) {
                count++;
            }
        }
        return count;
    }

    public RecordEncoder(int schemaVersion, List<DingoSchema> schemas, long id) {
        this.schemaVersion = schemaVersion;
        this.schemas = schemas;
        this.id = id;
        this.codecVersion = Config.CODEC_VERSION_V2;

        if (this.codecVersion == Config.CODEC_VERSION_V2) {
            int[] size = Utils.getApproPerRecordSizeV2(schemas);
            this.keyBufSize = size[0];
            this.valueBufSize = size[1];
        } else {
            int[] size = Utils.getApproPerRecordSize(schemas);
            this.keyBufSize = size[0];
            this.valueBufSize = size[1];
        }
        this.valueColumnCount = getValueColumnCount();
    }

    public RecordEncoder(int schemaVersion, long id) {
        this.schemaVersion = schemaVersion;
        this.id = id;
        this.codecVersion = Config.CODEC_VERSION_V2;
    }

    public RecordEncoder(int codecVersion, int schemaVersion, List<DingoSchema> schemas, long id) {
        this.schemaVersion = schemaVersion;
        this.schemas = schemas;
        this.id = id;
        this.codecVersion = codecVersion;

        if (this.codecVersion == Config.CODEC_VERSION_V2) {
            int[] size = Utils.getApproPerRecordSizeV2(schemas);
            this.keyBufSize = size[0];
            this.valueBufSize = size[1];
        } else {
            int[] size = Utils.getApproPerRecordSize(schemas);
            this.keyBufSize = size[0];
            this.valueBufSize = size[1];
        }

        this.valueColumnCount = getValueColumnCount();
    }

    public RecordEncoder(int codecVersion, int schemaVersion, long id) {
        this.schemaVersion = schemaVersion;
        this.id = id;
        this.codecVersion = codecVersion;
    }

    private void encodePrefix(Buf buf) {
        buf.write((byte) 'r');
        buf.writeLong(id);
    }

    private void encodeTag(Buf buf) {
        buf.reverseWrite((byte)this.codecVersion);
        buf.reverseWrite((byte) 0);
        buf.reverseWrite((byte) 0);
        buf.reverseWrite((byte) 0);
    }

    private void encodeReverseTag(Buf buf) {
        buf.reverseWrite(Config.CODEC_VERSION);
        buf.reverseWrite((byte) 0);
        buf.reverseWrite((byte) 0);
        buf.reverseWrite((byte) 0);
    }

    private void encodeSchemaVersion(Buf buf) {
        buf.writeInt(schemaVersion);
    }

    private void encodeSchemaVersion(int pos, Buf buf) {
        buf.writeInt(pos, schemaVersion);
    }

    public KeyValue encode(Object[] record) {
        KeyValue kv = new KeyValue(null, null);
        kv.setKey(encodeKey(record));
        kv.setValue(encodeValue(record));
        return kv;
    }

    private byte[] encodeKeyV1(Object[] record) {
        Buf keyBuf = new BufImpl(keyBufSize);

        encodeTag(keyBuf);
        encodePrefix(keyBuf);

        for (DingoSchema schema : schemas) {
            if (schema.isKey()) {
                schema.encodeKey(keyBuf, record[schema.getIndex()]);
            }
        }
        return keyBuf.getBytes();
    }

    private byte[] encodeKeyV2(Object[] record) {
        Buf keyBuf = new BufImpl(keyBufSize);

        encodeTag(keyBuf);
        encodePrefix(keyBuf);

        for (DingoSchema schema : schemas) {
            if (schema.isKey()) {
                schema.encodeKeyV2(keyBuf, record[schema.getIndex()]);
            }
        }
        return keyBuf.getBytes();
    }

    public byte[] encodeKey(Object[] record) {
        if ( this.codecVersion == Config.CODEC_VERSION) {
            return encodeKeyV1(record);
        } else {
            return encodeKeyV2(record);
        }
    }

    public byte[] encodeValue(Object[] record) {
        if ( this.codecVersion == Config.CODEC_VERSION) {
            return encodeValueV1(record);
        } else {
            return encodeValueV2(record);
        }
    }

    public byte[] encodeMinKeyPrefix() {
        Buf keyBuf = new BufImpl(9);
        encodePrefix(keyBuf);
        return keyBuf.getBytes();
    }

    public byte[] encodeMaxKeyPrefix() {
        if (id == Long.MAX_VALUE) {
            throw new RuntimeException("CommonId reach max! Cannot generate Max Key Prefix");
        }
        Buf keyBuf = new BufImpl(9);
        keyBuf.write((byte) 'r');
        keyBuf.writeLong(id + 1);
        return keyBuf.getBytes();
    }

    public byte[] encodeKeyPrefix(Object[] record, int columnCount) {
        Buf keyBuf = new BufImpl(keyBufSize);
        encodePrefix(keyBuf);
        for (DingoSchema schema : schemas) {
            if (schema.isKey()) {
                if (columnCount-- > 0) {
                    schema.encodeKeyPrefix(keyBuf, record[schema.getIndex()]);
                } else {
                    break;
                }
            }
        }
        return keyBuf.getBytes();
    }

    public byte[] updateValueByRecord(byte[] buf, Object[] record, int[] columnIndexes) {
        Buf valueBuf = new BufImpl(buf);
        if (valueBuf.readInt() != schemaVersion) {
            throw new RuntimeException("Wrong Schema Version");
        }
        for (DingoSchema schema : schemas) {
            if (!schema.isKey()) {
                if (Arrays.binarySearch(columnIndexes, schema.getIndex()) < 0) {
                    schema.skipValue(valueBuf);
                } else {
                    schema.encodeValue(valueBuf, record[schema.getIndex()]);
                }
            }
        }
        return valueBuf.getBytes();
    }

    public byte[] resetKeyPrefix(byte[] key, long prefix) {
        BufImpl buf = new BufImpl(key);
        buf.skip(1);
        buf.writeLong(prefix);
        return key;
    }

    public byte[] updateValueByColumns(byte[] buf, Object[] updateColumn, int[] columnIndexes) {
        return null;
    }
}
